File: src/stores/groups.ts
import { defineStore } from "pinia";
import { computed, ref } from "vue";
import { api, type ProviderGroup, type GroupWithMembers } from "../api";
import { useAsyncOperation } from "../composables/useAsyncOperation";
import { useSwrCache } from "../composables/useSwrCache";
import { notifyDataChanged } from "../composables/useCrossStoreNotify";

export const useGroupsStore = defineStore("groups", () => {
  // SWR 缓存：30 秒 TTL
  const cache = useSwrCache<ProviderGroup[]>(30_000);
  const crudOp = useAsyncOperation();

  const groups = computed(() => cache.data.value ?? []);
  const currentGroup = ref<GroupWithMembers | null>(null);
  const loading = computed(() => cache.loading.value);
  const error = computed(() => cache.error.value);

  async function fetchAll() {
    return cache.fetchIfNeeded(() => api.listGroups());
  }

  async function fetchById(id: string) {
    const result = await crudOp.run(() => api.getGroup(id), "获取分组失败");
    if (result) currentGroup.value = result;
  }

  async function create(data: { name: string; model_name: string; strategy?: string }) {
    const result = await crudOp.runOrThrow(() => api.createGroup(data), "创建失败");
    if (result) {
      const list = cache.data.value ?? [];
      list.unshift(result);
      cache.data.value = list;
      notifyDataChanged("groups");
    }
    return result;
  }

  async function update(id: string, data: Partial<ProviderGroup>) {
    const result = await crudOp.runOrThrow(() => api.updateGroup(id, data), "更新失败");
    if (result) {
      const list = cache.data.value;
      if (list) {
        const idx = list.findIndex((g) => g.id === id);
        if (idx >= 0) list[idx] = result;
      }
      notifyDataChanged("groups");
    }
    return result;
  }

  async function remove(id: string) {
    await crudOp.runOrThrow(() => api.deleteGroup(id), "删除失败");
    const list = cache.data.value;
    if (list) {
      cache.data.value = list.filter((g) => g.id !== id);
      notifyDataChanged("groups");
    }
  }

  async function addMember(groupId: string, providerId: string, weight?: number) {
    await crudOp.runOrThrow(
      () => api.addGroupMember(groupId, { provider_id: providerId, weight }),
      "添加成员失败"
    );
    await fetchById(groupId);
  }

  async function removeMember(id: string, groupId: string) {
    await crudOp.runOrThrow(() => api.removeGroupMember(id), "移除成员失败");
    await fetchById(groupId);
  }

  /** 强制刷新（忽略缓存） */
  async function refresh() {
    return cache.refresh(() => api.listGroups());
  }

  return {
    groups,
    currentGroup,
    loading,
    error,
    crudLoading: crudOp.loading,
    crudError: crudOp.error,
    fetchAll,
    fetchById,
    create,
    update,
    remove,
    addMember,
    removeMember,
    refresh,
  };
});

File: src-tauri/src/commands/groups.rs
use tauri::State;

use crate::application::group_service::{
    self, AddMemberPayload, CreateGroupPayload, GroupWithMembersResponse, UpdateGroupPayload,
    UpdateMemberPayload,
};
use crate::models::{GroupMember, ProviderGroup};
use crate::AppState;

#[tauri::command]
pub async fn list_groups(_state: State<'_, AppState>) -> Result<Vec<ProviderGroup>, String> {
    group_service::list(_state).await
}

#[tauri::command]
pub async fn find_groups_by_model(
    state: State<'_, AppState>,
    model_name: String,
) -> Result<Vec<ProviderGroup>, String> {
    group_service::find_by_model(state, model_name).await
}

#[tauri::command]
pub async fn get_group(
    state: State<'_, AppState>,
    id: String,
) -> Result<GroupWithMembersResponse, String> {
    group_service::get(state, id).await
}

#[tauri::command]
pub async fn create_group(
    state: State<'_, AppState>,
    payload: CreateGroupPayload,
) -> Result<ProviderGroup, String> {
    group_service::create(state, payload).await
}

#[tauri::command]
pub async fn update_group(
    state: State<'_, AppState>,
    id: String,
    payload: UpdateGroupPayload,
) -> Result<ProviderGroup, String> {
    group_service::update(state, id, payload).await
}

#[tauri::command]
pub async fn delete_group(state: State<'_, AppState>, id: String) -> Result<bool, String> {
    group_service::delete(state, id).await
}

#[tauri::command]
pub async fn add_group_member(
    state: State<'_, AppState>,
    group_id: String,
    payload: AddMemberPayload,
) -> Result<GroupMember, String> {
    group_service::add_member(state, group_id, payload).await
}

#[tauri::command]
pub async fn update_group_member(
    state: State<'_, AppState>,
    id: String,
    payload: UpdateMemberPayload,
) -> Result<GroupMember, String> {
    group_service::update_member(state, id, payload).await
}

#[tauri::command]
pub async fn remove_group_member(state: State<'_, AppState>, id: String) -> Result<bool, String> {
    group_service::remove_member(state, id).await
}
